File: varmor/varmor.go
// Package varmor provides versioned armoring for arbitrary sequences of bytes.
//
// The armored form is free of whitespace (including newlines), safe to embed in URLs (other than possibly
// its length) and safe to pass unescaped in a POSIX shell.
package varmor

import (
	"encoding/base64"
	"errors"
	"fmt"
	"strings"
)

const (
	magicPrefix = "saltybox"
	v1Magic     = "saltybox1:"
)

// Wrap an array of bytes in armor, returning the resulting string.
func Wrap(body []byte) string {
	encoded := base64.RawURLEncoding.EncodeToString(body)

	return fmt.Sprintf("%s%s", v1Magic, encoded)
}

// Unwrap an armored string.
//
// Error conditions include:
//
//   - The input is provably truncated.
//   - Base64 decoding failure.
//   - Input indicates a future version of the format that we do not support.
//   - Input does not appear to be the result of Wrap().
func Unwrap(varmoredBody string) ([]byte, error) {
	if len(varmoredBody) < len(v1Magic) {
		return nil, errors.New("input size smaller than magic marker; likely truncated")
	}

	switch {
	case strings.HasPrefix(varmoredBody, v1Magic):
		armoredBody := strings.TrimPrefix(varmoredBody, v1Magic)
		body, err := base64.RawURLEncoding.DecodeString(armoredBody)
		if err != nil {
			return nil, fmt.Errorf("base64 decoding failed: %w", err)
		}
		return body, nil
	case strings.HasPrefix(varmoredBody, magicPrefix):
		return nil, errors.New("input claims to be saltybox, but not a version we support")
	default:
		return nil, errors.New("input unrecognized as saltybox data")
	}
}

File: saltybox-rs/src/varmor.rs
use base64::{engine::general_purpose::URL_SAFE_NO_PAD, Engine as _};

const MAGIC_PREFIX: &str = "saltybox";
const V1_MAGIC: &str = "saltybox1:";

pub fn wrap(body: &[u8]) -> String {
    let encoded = URL_SAFE_NO_PAD.encode(body);
    format!("{}{}", V1_MAGIC, encoded)
}

#[derive(thiserror::Error, Debug)]
pub enum UnwrapError {
    #[error("input size smaller than magic marker; likely truncated")]
    Truncated,
    #[error("base64 decoding failed: {0}")]
    Base64(#[from] base64::DecodeError),
    #[error("input claims to be saltybox, but not a version we support")]
    WrongVersion,
    #[error("input unrecognized as saltybox data")]
    Unrecognized,
}

pub fn unwrap(varmored_body: &str) -> Result<Vec<u8>, UnwrapError> {
    if varmored_body.len() < V1_MAGIC.len() {
        return Err(UnwrapError::Truncated);
    }

    if let Some(armored_body) = varmored_body.strip_prefix(V1_MAGIC) {
        let body = URL_SAFE_NO_PAD.decode(armored_body)?;
        return Ok(body);
    }

    if varmored_body.starts_with(MAGIC_PREFIX) {
        return Err(UnwrapError::WrongVersion);
    }

    Err(UnwrapError::Unrecognized)
}

